File: Substring With Concatenation of All Words.py
# Leetcode Problem 30: Substring With Concatenation of All Words
# PYTHON CODE
from collections import Counter

class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        if not s or not words:
            return []

        word_len = len(words[0])
        word_count = len(words)
        total_len = word_len * word_count

        target = Counter(words)
        ans = []

        for offset in range(word_len):
            left = offset
            curr = Counter()
            count = 0

            for right in range(offset, len(s) - word_len + 1, word_len):
                word = s[right:right + word_len]

                if word in target:
                    curr[word] += 1
                    count += 1

                    while curr[word] > target[word]:
                        left_word = s[left:left + word_len]
                        curr[left_word] -= 1
                        count -= 1
                        left += word_len

                    if count == word_count:
                        ans.append(left)

                        left_word = s[left:left + word_len]
                        curr[left_word] -= 1
                        count -= 1
                        left += word_len
                else:
                    curr.clear()
                    count = 0
                    left = right + word_len

        return ans

File: Substring With Concatenation of All Words.cpp
// Leetcode Problem 30: Substring With Concatenation of All Words
// C++ CODE
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        vector<int> ans;

        if (s.empty() || words.empty())
            return ans;

        int wordLen = words[0].size();
        int wordCount = words.size();

        unordered_map<string, int> target;
        for (string &word : words)
            target[word]++;

        for (int offset = 0; offset < wordLen; offset++) {
            int left = offset;
            int count = 0;
            unordered_map<string, int> window;

            for (int right = offset; right + wordLen <= s.size(); right += wordLen) {
                string word = s.substr(right, wordLen);

                if (target.count(word)) {
                    window[word]++;
                    count++;

                    while (window[word] > target[word]) {
                        string leftWord = s.substr(left, wordLen);
                        window[leftWord]--;
                        count--;
                        left += wordLen;
                    }

                    if (count == wordCount) {
                        ans.push_back(left);

                        string leftWord = s.substr(left, wordLen);
                        window[leftWord]--;
                        count--;
                        left += wordLen;
                    }
                } else {
                    window.clear();
                    count = 0;
                    left = right + wordLen;
                }
            }
        }

        return ans;
    }
};